HeterogeneousCore/AlpakaCore/interface/alpaka/ESProducer.h

```diff
@@ -111,12 +115,28 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
               } else {
                 anynull = true;
               }
-              queues.push_back(deviceRecord.queuePtr());
-            }
-            // TODO: to be changed asynchronous later
-            for (auto& queuePtr : queues) {
-              alpaka::wait(*queuePtr);
+              if (synchronize) {
+                alpaka::wait(deviceRecord.queue());
+              } else {
+                enqueueCallback(deviceRecord.queue(), std::move(holder));
+              }
+              // The Queue is returned to the QueueCache. The same
+              // Queue may be used to for other work before the work
+              // enqueued here finishes. The only impact would be a
+              // (slight?) delay in the completion of the other work.
+              // Given that the ESProducers are expected to be mostly
+              // for host-to-device data copies, that are serialized
+              // anyway (at least on current NVIDIA), this should be
+              // reasonable behavior for now.
```

Comment:
Alternatively the Queues could be stored in an `std::vector` like before, that is then moved from the "acquire" lambda to be "produce" lambda.

Reply:
> The `Queue` is returned to the `QueueCache`. The same `Queue` may be used to for other work before the work enqueued here finishes.

Isn't this already a common behaviour with queues in the `QueueCache` ?
Or, in the other cases, queues are returned only once all work has been completed ?

Reply:
> > The `Queue` is returned to the `QueueCache`. The same `Queue` may be used for other work before the work enqueued here finishes.
> 
> Isn't this already a common behaviour with queues in the `QueueCache` ? Or, in the other cases, queues are returned only once all work has been completed ?

In the Event system case the queues are returned only after all work has been completed. This happens because the `shared_ptr<Queue>` are held in the Event data products, and they return the queue only after synchronization. (although I'd say this behavior is more of a result of other design decisions than being a deliberate choice by itself)

Reply:
Ah, yes. It's the `EventCache` that has additional provisions to avoid reusing events that are not yet "ready".

IMHO these current changes are fine, because Event Setup transitions are supposed to be "rare" with respect to the Event work.
If we later notice any performance impact here, we can always add back the `vector`, or extend the `QueueCache` to only return queues that are "empty".